feat(polys): make FiniteExtension a proper Domain

oscarbenjamin opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

See discussion in #20608.

#### Brief description of what is fixed or changed

Makes MonogenicFiniteExtension into a proper domain that can be used with Poly etc:
```python
In [1]: from sympy.polys.agca.extensions import FiniteExtension

In [2]: K = FiniteExtension(Poly(x**2 - 2))

In [3]: K
Out[3]: ZZ[x]/(x**2 - 2)

In [4]: p = Poly(x + y, y, domain=K)

In [5]: p
Out[5]: Poly(y + x, y, domain='ZZ[x]/(x**2 - 2)')

In [6]: p**2
Out[6]: Poly(y**2 + 2*x*y + 2, y, domain='ZZ[x]/(x**2 - 2)')
```
This also makes it possible to build a chain of extensions:
```python
In [7]: K2 = FiniteExtension(Poly(y**2 - 3, domain=K))

In [8]: K2
Out[8]: ZZ[x]/(x**2 - 2)[y]/(y**2 - 3)

In [9]: xf = K2.convert(x)

In [10]: yf = K2.convert(y)

In [11]: xf + yf
Out[11]: x + y

In [12]: _ ** 2
Out[12]: 2*x*y + 5
```
I've added a `Domain.drop` method to drop a generator:
```python
In [13]: QQ[x, y].drop(x)
Out[13]: QQ[y]

In [14]: QQ[x, y][z].drop(x)
Out[14]: QQ[y][z]

In [15]: QQ[x, y][z].get_field()
Out[15]: QQ[x,y](z)

In [16]: QQ[x, y][z].get_field().drop(x)
Out[16]: QQ[y](z)

In [17]: QQ[x, y].get_field()[z]
Out[17]: QQ(x,y)[z]

In [18]: QQ[x, y].get_field()[z].drop(x)
Out[18]: QQ(y)[z]

In [19]: QQ[x, y].get_field()[z].drop(x, y)
Out[19]: QQ[z]
```

#### Other comments

It seems that FiniteExtension uses `old_poly_ring`. I'm not sure if that's something that should be changed. I guess this should really be called `dense_poly_ring` and really it does make sense to use a dense representation here since the polys are limited in size by the modulus. There are however some bugs in conversion between old and new poly ring and also between different poly rings.


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* polys
    * A new polys FiniteExtension domain is added.
<!-- END RELEASE NOTES -->

oscarbenjamin committed:
feat(polys): make FiniteExtension a proper Domain

oscarbenjamin commented:
This should work but doesn't:
```python
In [34]: K1 = FiniteExtension(Poly(x**2 - 2))

In [35]: K2 = FiniteExtension(Poly(y**3 - 2))

In [36]: p1 = Poly(x + t, t, domain=K1)

In [37]: p2 = Poly(y + z, z, domain=K2)

In [40]: p1
Out[40]: Poly(t + x, t, domain='ZZ[x]/(x**2 - 2)')

In [41]: p2
Out[41]: Poly(z + y, z, domain='ZZ[y]/(y**3 - 2)')

In [42]: p1 * p2
---------------------------------------------------------------------------
GeneratorsError  
```
The problem is a bug I think in old poly ring:
```python
In [44]: K1 = QQ.old_poly_ring(x)

In [45]: K2 = K1.old_poly_ring(y)

In [46]: K1
Out[46]: QQ[x]

In [47]: K2
Out[47]: QQ[x][y]

In [49]: K2.convert(K1.convert(x), K1)
---------------------------------------------------------------------------
GeneratorsError                           Traceback (most recent call last)
<ipython-input-49-7f560bdf4669> in <module>
----> 1 K2.convert(K1.convert(x), K1)

~/current/sympy/sympy/sympy/polys/domains/domain.py in convert(self, element, base)
    111 
    112         if base is not None:
--> 113             return self.convert_from(element, base)
    114 
    115         if self.of_type(element):

~/current/sympy/sympy/sympy/polys/domains/domain.py in convert_from(self, element, base)
     98 
     99         if _convert is not None:
--> 100             result = _convert(element, base)
    101 
    102             if result is not None:

~/current/sympy/sympy/sympy/polys/domains/old_polynomialring.py in from_GlobalPolynomialRing(K1, a, K0)
    116                 return K1(a.convert(K1.dom).rep)
    117         else:
--> 118             monoms, coeffs = _dict_reorder(a.to_dict(), K0.gens, K1.gens)
    119 
    120             if K1.dom != K0.dom:

~/current/sympy/sympy/sympy/polys/polyutils.py in _dict_reorder(rep, gens, new_gens)
    416             for monom in monoms:
    417                 if monom[i]:
--> 418                     raise GeneratorsError("unable to drop generators")
    419 
    420     return map(tuple, new_monoms), coeffs

GeneratorsError: unable to drop generators
```

oscarbenjamin committed:
feat(polys): make FiniteExtension a proper Domain

oscarbenjamin committed:
feat(polys): make FiniteExtension a proper Domain

Add tests for FiniteExtension

oscarbenjamin commented:
The sphinx build is failing with:
```
<unknown>:1: WARNING: py:class reference target not found: ExtensionElement
```
I have tried but I'm not suer how to debug this further. I've grepped for ExtensionElement in the diff and the codebase and there are not many mentions and I don't see any of relevance. I've also tried running `sphinx-build` with `-vvv` but it didn't give any more useful information.

I'm at a loss...

oscarbenjamin committed:
feat(polys): make FiniteExtension a proper Domain

Add ExtensionElement to rst docs

sylee957 commented:
`convert_from` does not work
```
from sympy import *
from sympy.polys.agca.extensions import FiniteExtension

x = Symbol('x')
p = Poly([1, 0, 1], x, domain=QQ)
field = FiniteExtension(p)
field.convert_from(QQ(1, 2), QQ)
```

oscarbenjamin committed:
feat(polys): make FiniteExtension a proper domain

Fix a few docstrings

oscarbenjamin committed:
feat(polys): make FiniteExtension a proper domain

Fix FiniteExtension.convert_from

oscarbenjamin commented:
> `convert_from` does not work

I've fixed that.

This probably needs more tests e.g. for Poly.

@jksuom @sylee957 what do you think of this otherwise?

jksuom commented on sympy/polys/agca/extensions.py:
```diff
@@ -67,7 +73,14 @@ def __rsub__(f, g):
+    def __mod__(f, g):
+        rep = f._get_rep(g)
+        if rep is not None:
+            return ExtElem((f.rep % rep) % f.ext.mod, f.ext)
```
I don't think that `__mod__` is well defined in this class. `f.rep` is usually the canonical representative of lowest degree but the formula should work for all representatives. So `((f.rep + f.ext.mod) % rep)  % f.ext.mod` should be the same as `(f.rep % rep) % f.ext.mod`. This is generally true only if `f.ext.mod % rep` is zero, or `rep` is a factor of `f.ext.mod`.

Will this be needed somewhere?

oscarbenjamin replied:
It is needed when constructing a nested extension. If I remove this method then run this:
```python
from sympy import *
from sympy.polys.agca.extensions import FiniteExtension
from sympy.abc import x, y

Kx = FiniteExtension(Poly(x**2 - 2, x, domain=ZZ))
Ky = FiniteExtension(Poly(y**2 - 2, y, domain=ZZ))
Kxy = FiniteExtension(Poly(y**2 - 2, y, domain=Kx))
```
That gives
```console
$ python t.py 
Traceback (most recent call last):
  File "t.py", line 7, in <module>
    Kxy = FiniteExtension(Poly(y**2 - 2, y, domain=Kx))
  File "/Users/enojb/current/sympy/sympy/sympy/polys/agca/extensions.py", line 222, in __init__
    mod, rem = mod.div(mod.LC())
  File "/Users/enojb/current/sympy/sympy/sympy/polys/polytools.py", line 1680, in div
    q, r = F.div(G)
  File "/Users/enojb/current/sympy/sympy/sympy/polys/polyclasses.py", line 491, in div
    q, r = dmp_div(F, G, lev, dom)
  File "/Users/enojb/current/sympy/sympy/sympy/polys/densearith.py", line 1629, in dmp_div
    return dmp_rr_div(f, g, u, K)
  File "/Users/enojb/current/sympy/sympy/sympy/polys/densearith.py", line 1372, in dmp_rr_div
    return dup_rr_div(f, g, K)
  File "/Users/enojb/current/sympy/sympy/sympy/polys/densearith.py", line 1337, in dup_rr_div
    if lc_r % lc_g:
TypeError: unsupported operand type(s) for %: 'ExtensionElement' and 'ExtensionElement'
```
Maybe there's a better way.

oscarbenjamin replied:
Perhaps mod should only be defined for constants.

jksuom replied:
It looks like this could work for constants.

jksuom replied:
Invertible constants, and the result would be zero.

oscarbenjamin replied:
I've changed it so that `f % g` raises `NotImplementedError` unless `g.is_ground`.

oscarbenjamin committed:
feat(polys): make FiniteExtension a proper domain

Only allow mod over the ground domain because it is not well defined in
general for FiniteExtension.

oscarbenjamin commented on sympy/polys/agca/extensions.py:
```diff
@@ -224,22 +240,53 @@ def __init__(self, mod):
+
+    def exquo(self, f, g):
+        rep = self.domain.exquo(f.rep, g.rep)
+        return ExtElem(rep % self.mod, self)
```
I'm not sure how to implement `exquo` here.

If self is `QQ[x]/(x**2 - 2)` then `self.domain` is `QQ`. At least this should call `self.ring.exquo` rather than `self.domain.exquo` (since the arguments are elements of `self.ring` so that's just a mistake). However `self.ring` is a ring whereas `self` is a field in which exact division should be possible for any nonzero divisor.

jksuom replied:
Shouldn't this be the same as `__truediv__`? So this could be `return f / g`.

oscarbenjamin replied:
If `f / g` is exact then does that not imply that `f // g == f / g` and `f % g == 0`?

jksuom replied:
It looks like that is how `f // g` and `f % g` should be implemented when `ring.domain` is a field. Otherwise some kind of error should probably be returned.

oscarbenjamin replied:
In a nested extension the base domain is also a finite extension:
```python
In [17]: from sympy.polys.agca.extensions import FiniteExtension

In [18]: K = FiniteExtension(Poly(x**2 - 2, field=True))

In [19]: K2 = FiniteExtension(Poly(y**2 - 2, domain=K))

In [20]: K2
Out[20]: QQ[x]/(x**2 - 2)[y]/(y**2 - 2)

In [21]: K2.domain
Out[21]: QQ[x]/(x**2 - 2)
```
We need to be able to determine whether or not this is a field.

Can we say that if `K` is a field and `p(x)` is irreducible in `K[x]` then `K[x]/p(x)` is a field?

Conversely can we say that if `K` is not a field or `p(x)` is reducible over `K[x]` then `K[x]/p(x)` is not a field?

It seems like we can do
```
    self.is_Field = domain.is_field and mod.is_irreducible
```
in `Domain.__init__`.

Then if `is_Field` is true we can allow full division with `/`, `//` and `%` as:
```
f / g  ==  f * g.invert(mod)
f // g == f / g
f % g == 0
```
Otherwise if `is_Field` is false then we can allow the same but only if g is in K (`g.is_ground`)?

In non-field bug ground case the base domain `K` can decide if the division should succeed or fail.

jksuom replied:
It looks like `is_Field` currently means that "exact" division (by nonzerodivisors) is possible. It is true for RR though it is not really a field (only few elements are actually invertible). I would consider making `is_Field` true for `K[x]/p(x)` whenever `K` is true even if `p(x)` is not irreducible. If there turn out to be issues, then maybe `is_Field` could be None instead of false.

If `K` is not a field, then `is_Field` should be false. In that case, I think that it is safe to divide only by invertible constants g in K.

oscarbenjamin replied:
Okay, I'll try that.

It seems that the `is_unit` method is broken for `ZZ` and `QQ[x]` unless I misunderstand:
```python
In [62]: ZZ.is_unit(ZZ(-1))
Out[62]: False

In [63]: K = QQ[x]

In [64]: K.is_unit(K(2))
Out[64]: False

In [66]: QQ.is_unit(QQ(2))
Out[66]: True

In [68]: QQ.is_unit(QQ(0))
Out[68]: False
```

oscarbenjamin committed:
feat(polys): make FiniteExtension a proper domain

Allow division of an ExtensionElement over a ring but only where the
divisor is an invertible constant (unit) of the ground domain.

For example in QQ[y][x]/(x**2 - 2) we can divide by any nonzero element
of QQ (the units of QQ[y]). In ZZ[x]/(x**2 - 2) we can only divide by 1
or -1.

If the ground domain for a FiniteExtension is a field then it is assumed
that the extension is also a field although this might not hold if the
quotient polynomial is not irreducible (this is not checked). If the
ground domain is a field then division by any nonzero element is
allowed.

This applies in particular to nested extensions e.g.:
  K = QQ[x]/(x**2 - 2)[y]/(y**2 - x)
Here the ground domain is QQ[x]/(x**2 - 2) which is assumed to be a
field because QQ is. It is therefore assumed that K is a field as well.
Division by any nonzero element of K is then permitted and considered
exact so that f % g is always zero.

oscarbenjamin committed:
feat(polys): make FiniteExtension a proper domain

Fix the Domain.is_unit method for rings.

oscarbenjamin committed:
feat(polys): make FiniteExtension a proper domain

Add tests for FiniteExtension

oscarbenjamin commented:
@jksuom I tried replacing `old_poly_ring` with `poly_ring` in `FiniteExtension` and it seems to be mostly working but I can't find the equivalent of `DMP.invert` for `PolyElement`.
```
In [9]: Poly(x, x, domain=QQ).invert(Poly(x**2 - 2, x, domain=QQ))
Out[9]: Poly(1/2*x, x, domain='QQ')
```
Is there a way to do this with `PolyElement`?

jksuom commented:
> Is there a way to do this with PolyElement?

I don't think that has been implemented. But why would you want to use sparse polynomials?

oscarbenjamin commented:
> But why would you want to use sparse polynomials?

I'm wondering if it could make a more efficient implementation of nested extensions. Basically I'm thinking about being able to use domain matrix to work with a system of equations like this one:
https://stackoverflow.com/questions/65293565/performance-issues-with-solve-in-sympy?noredirect=1#comment115437481_65293565

The domain of the matrix elements is:
```
QQ[s1,s2,s3,c1,c2,c3,l1,l2,l3,m1,m2,m3]
```
if we treat e.g. `sin(theta1)` and `cos(theta1)` as generators `s1` and `c1` in a polynomial ring. However this is not correct because `s1` and `c1` are separately transcendental but not algebraically independent over `QQ`. The domain should be something like
```
QQ<s1,s2,s3,c1,c2,c3>[l1,l2,l3,m1,m2,m3]
```
where `QQ<s1,s2,s3,c1,c2,c3>` could be implemented here as
```
QQ[s1][c1]/(s1**2 + c1**2 - 1)[s2][c2]/(s2**2 + c2**2 - 1)[s3][c3]/(s3**2 + c3**2 - 1)
```
But now this is like the dense multivariate implementation except worse because we have a DMP which has a list of DMP etc rather than a single DMP with a list of lists of lists. Most of the DMP will be in trivial zero or constant form like `[]` or `[something]`. Perhaps that's okay if all extensions have low degree but it does mean that trivial object like `1` ends up with a complicated representation like
```
DMP([DMP([DMP([DMP([DMP([DMP([1], QQ, QQ[c1])], QQ[c1], QQ[c1][s1])], QQ[c1][s1], QQ[c1][s1][c2])], QQ[c1][s1][c2], QQ[c1][s1][c2][s2])], QQ[c1][s1][c2][s2], QQ[c1][s1][c2][s2][c3])], QQ[c1][s1][c2][s2][c3], QQ[c1][s1][c2][s2][c3][s3])
```

I was wondering if it would be possible to represent deeply nested extensions using a flattened `PolyElement` representation somehow that could be more efficient. Maybe in some cases it would be necessary to literally nest the extensions e.g.:
```
QQ[x][y]/(y**2 - x)[z]/(z**2 - y)
```
However I think that there is a common case where we have many algebraically independent extensions like `cos(theta1),sin(theta1)`, `cos(theta2),sin(theta2)` etc so it should be possible to flatten the representation for that.

Maybe that isn't a good reason to use `PolyElement`. The other reason I tried it was because old poly rings don't work with nesting:
```
In [15]: Kx = QQ.old_poly_ring(x)

In [16]: Kxy = Kx.old_poly_ring(y)

In [17]: Kx
Out[17]: QQ[x]

In [18]: Kxy
Out[18]: QQ[x][y]

In [19]: Kxy.convert(Kx.convert(1), Kx)
Out[19]: DMP([DMP([1], QQ, QQ[x])], QQ[x], QQ[x][y])

In [20]: Kxy.convert(Kx.convert(x), Kx)
---------------------------------------------------------------------------
GeneratorsError 
```

jksuom commented:
I think that `invert` for univariate sparse polynomials could be easily implemented by means of `half_gcdex`. (It will pass to dense polynomials.)

jksuom commented:
I think that consecutive finite extensions should be combined into a single finite extension of higher dimension in the same way as algebraic number fields are combined.

oscarbenjamin commented:
> I think that consecutive finite extensions should be combined into a single finite extension of higher dimension in the same way as algebraic number fields are combined.

For algebraically independent extensions I think that will make the arithmetic much more complicated than it needs to be:
```
In [39]: minpoly(sqrt(1-c1**2) + sqrt(1-c2**2) + sqrt(1 - c3**2))
Out[39]: 
 8    6 ⎛    2       2       2     ⎞    4 ⎛    4       2   2       2   2        2       4       2   2        2       4        2 
x  + x ⋅⎝4⋅c₁  + 4⋅c₂  + 4⋅c₃  - 12⎠ + x ⋅⎝6⋅c₁  + 4⋅c₁ ⋅c₂  + 4⋅c₁ ⋅c₃  - 20⋅c₁  + 6⋅c₂  + 4⋅c₂ ⋅c₃  - 20⋅c₂  + 6⋅c₃  - 20⋅c₃  

    ⎞    2 ⎛    6       4   2       4   2       4       2   4        2   2   2        2   2       2   4        2   2        2   
+ 30⎠ + x ⋅⎝4⋅c₁  - 4⋅c₁ ⋅c₂  - 4⋅c₁ ⋅c₃  - 4⋅c₁  - 4⋅c₁ ⋅c₂  + 40⋅c₁ ⋅c₂ ⋅c₃  - 24⋅c₁ ⋅c₂  - 4⋅c₁ ⋅c₃  - 24⋅c₁ ⋅c₃  + 28⋅c₁  + 

    6       4   2       4       2   4        2   2        2       6       4        2     ⎞     8       6   2       6   2       6
4⋅c₂  - 4⋅c₂ ⋅c₃  - 4⋅c₂  - 4⋅c₂ ⋅c₃  - 24⋅c₂ ⋅c₃  + 28⋅c₂  + 4⋅c₃  - 4⋅c₃  + 28⋅c₃  - 28⎠ + c₁  - 4⋅c₁ ⋅c₂  - 4⋅c₁ ⋅c₃  + 4⋅c₁ 

       4   4       4   2   2       4   2       4   4       4   2       4       2   6       2   4   2       2   4       2   2   4
 + 6⋅c₁ ⋅c₂  + 4⋅c₁ ⋅c₂ ⋅c₃  - 4⋅c₁ ⋅c₂  + 6⋅c₁ ⋅c₃  - 4⋅c₁ ⋅c₃  - 2⋅c₁  - 4⋅c₁ ⋅c₂  + 4⋅c₁ ⋅c₂ ⋅c₃  - 4⋅c₁ ⋅c₂  + 4⋅c₁ ⋅c₂ ⋅c₃ 

        2   2   2        2   2       2   6       2   4        2   2        2     8       6   2       6       4   4       4   2  
 - 24⋅c₁ ⋅c₂ ⋅c₃  + 20⋅c₁ ⋅c₂  - 4⋅c₁ ⋅c₃  - 4⋅c₁ ⋅c₃  + 20⋅c₁ ⋅c₃  - 12⋅c₁  + c₂  - 4⋅c₂ ⋅c₃  + 4⋅c₂  + 6⋅c₂ ⋅c₃  - 4⋅c₂ ⋅c₃  -

     4       2   6       2   4        2   2        2     8       6       4        2    
 2⋅c₂  - 4⋅c₂ ⋅c₃  - 4⋅c₂ ⋅c₃  + 20⋅c₂ ⋅c₃  - 12⋅c₂  + c₃  + 4⋅c₃  - 2⋅c₃  - 12⋅c₃  + 9
```
All that is really needed here is to reduce powers of `s1` with `s1**2 = 1 - c1**2` and likewise for `s2` and `s3`.

sylee957 commented:
I have an another question whether or not the FiniteExtension can allow non-monic modulus.
As we have decided that this can be called `Field` even though the zero divisors occur, I think that we can also loose other restrictions like monicity.

jksuom commented:
> whether or not the FiniteExtension can allow non-monic modulus.

It is essential that the modulus is monic or can be made monic by dividing by the leading coefficient. Otherwise the extension will not be finite. Consider, for example, the modulus `2*x - 1` over the integers. Any ring containing its root `1/2` will also contain all powers `1/2**n` and its elements cannot therefore be expressed as linear combinations with integer coefficients of any finite number of elements. In case of such moduli, the base domain should first be extended to include an inverse of the leading coefficient.

jksuom commented:
> All that is really needed here is to reduce powers of s1 with s1**2 = 1 - c1**2 and likewise for s2 and s3.

That would be essentially the tensor product of the three quadratic extensions generated by `s1`, `s2` and `s3`. Such a product is also a finite free extension with a basis consisting of low degree monomials in the generators. It will be isomorphic to the finite extension generated by a primitive element `t` (root of Out[39]).In fact, the two extensions can have the same elements. Only their bases are different, monomial basis instead of power basis..

The tensor product could be implemented by a class storing

* finite extension generated by a primitive element,
* expressions of the given generators as extension elements,
* basis change matrix from the power basis to the monomial basis.

This data should make it possible to define conversions `to_sympy` and `from_sympy`.

oscarbenjamin committed:
feat(polys): make FiniteExtension a proper domain

Add missing methods needed to construct and use nested finite
extensions.

oscarbenjamin committed:
feat(polys): make FiniteExtension a proper domain

Fix doctest

oscarbenjamin committed:
feat(polys): make FiniteExtension a proper domain

Add test of FiniteExtension for calculations with sin and cos.

oscarbenjamin commented:
@jksuom @sylee957 this is ready for review.

jksuom commented on sympy/polys/agca/extensions.py:
```diff
@@ -67,55 +73,94 @@ def __rsub__(f, g):
-        if not f.ext.domain.is_Field:
-            raise NotImplementedError("base field expected")
-        return ExtElem(f.rep.invert(f.ext.mod), f.ext)
+        f._divcheck()
```
Is there no need to check the return value.

oscarbenjamin replied:
The `_divcheck` function raises an exception unless the element can be inverted.

jksuom commented:
I'm currently unable to work with my development machine so I have only been reading the code but it looks good.

oscarbenjamin commented:
Thank for the review @jksuom. I'll merge this then because I think @sylee957 needed it for `eigenvects`.
